File: 프로그래머스/lv1/86491. 최소직사각형/최소직사각형.py
import sys
sys.setrecursionlimit(10**8)

msize = float('inf')
def solve(idx, sizes, h, w):
    if idx == len(sizes):
        global msize
        msize = min(msize, h * w)
        return
    if (sizes[idx][0] <= h and sizes[idx][1] <= w) or  (sizes[idx][1] <= h and sizes[idx][0] <= w):
        solve(idx + 1, sizes, h, w)
    else:
        solve(idx + 1, sizes, max(h, sizes[idx][1]), max(w, sizes[idx][0]))
        solve(idx + 1, sizes, max(h, sizes[idx][0]), max(w, sizes[idx][1]))

def solution(sizes):
    solve(0, sizes, sizes[0][0], sizes[0][1])
    global msize
    return msize

File: 프로그래머스/lv1/86491. 최소직사각형/최소직사각형.java
import java.util.*;

class Solution {
    long answer = 987654433211L;
    public void dfs(int idx, int now_w, int now_h, int[][] sizes){
        if(idx >= sizes.length){
            answer = Math.min(now_w * now_h, answer);
            return;
        }
        if((sizes[idx][0] <= now_w && sizes[idx][1] <= now_h) || (sizes[idx][1] <= now_w && sizes[idx][0] <= now_h)){
            dfs(idx + 1, now_w, now_h, sizes);
        }
         else{
             dfs(idx + 1, Math.max(sizes[idx][0], now_w), Math.max(sizes[idx][1], now_h), sizes);
            dfs(idx + 1, Math.max(sizes[idx][1], now_w), Math.max(sizes[idx][0], now_h), sizes);
         }
        return;
    }
    public long solution(int[][] sizes) {
        dfs(0, 0, 0, sizes);
        return answer;
    }
}